File: Exercism/dart/matching-brackets/lib/matching_brackets.dart

class MatchingBrackets {
  bool isPaired(String input) {
    List<String> brackets = [];

    for (int i = 0 ; i < input.length ; i++) {
      switch (input[i]) {
        case '{': { brackets.add(input[i]); }
        break;
        case '[': { brackets.add(input[i]); }
        break;
        case '(': { brackets.add(input[i]); }
        break;

        case '}': {
          if (brackets.isEmpty || brackets.last != '{')
            return false;
          else
            brackets.removeLast();
        }
        break;
        case ']': {
          if (brackets.isEmpty || brackets.last != '[')
            return false;
          else
            brackets.removeLast();
        }
        break;
        case ')': {
          if (brackets.isEmpty || brackets.last != '(')
            return false;
          else
            brackets.removeLast();
        }
        break;
      }
    }

    return brackets.isEmpty;
  }
}

File: Exercism/c/matching-brackets/src/matching_brackets.c
#include "matching_brackets.h"
#include <stdio.h>
#include <string.h>

char corresponding_bracket(char bracket);

bool is_paired(const char* input)
{
    int len = strlen(input);
    char stack[len];
    int idx = 0;

    for (int i = 0; i < len; ++i) {
        if (input[i] == '[' || input[i] == '{' || input[i] == '(') {
            stack[idx++] = input[i];
        } else if (input[i] == ']' || input[i] == '}' || input[i] == ')') {
            if (stack[idx - 1] != corresponding_bracket(input[i]))
                return false;
            idx--;
        }
    }

    return idx == 0;
}

char corresponding_bracket(char bracket)
{
    if (bracket == '[')
        return ']';
    if (bracket == ']')
        return '[';
    if (bracket == '(')
        return ')';
    if (bracket == ')')
        return '(';
    if (bracket == '}')
        return '{';
    if (bracket == '{')
        return '}';

    return ' ';
}